src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2960,41 +2969,37 @@ void ProcessAnalysisAnnotationsForMethod (MethodDefinition method, DependencyKin
-				break;
-			}
-
-			if (dependencyKind == DependencyKind.DynamicallyAccessedMember) {
-				// All override methods should have the same annotations as their base methods
-				// (else we will produce warning IL2046 or IL2092 or some other warning).
-				// When marking override methods via DynamicallyAccessedMembers, we should only issue a warning for the base method.
-				if (method.IsVirtual && Annotations.GetBaseMethods (method) != null)
-					return;
-			}
+			// All override methods should have the same annotations as their base methods
+			// (else we will produce warning IL2046 or IL2092 or some other warning).
+			// When marking override methods via DynamicallyAccessedMembers, we should only issue a warning for the base method.
+			bool skipWarningsForOverride = dependencyKind == DependencyKind.DynamicallyAccessedMember && method.IsVirtual && Annotations.GetBaseMethods (method) != null;
 
-			if (Annotations.DoesMethodRequireUnreferencedCode (method, out RequiresUnreferencedCodeAttribute? requiresUnreferencedCode))
-				ReportRequiresUnreferencedCode (method.GetDisplayName (), requiresUnreferencedCode, new DiagnosticContext (origin, diagnosticsEnabled: true, Context));
+			bool warnsOnRUC = Annotations.DoesMethodRequireUnreferencedCode (method, out RequiresUnreferencedCodeAttribute? requiresUnreferencedCode);
+			if (warnsOnRUC && !skipWarningsForOverride)
+				ReportRequiresUnreferencedCode (method.GetDisplayName (), requiresUnreferencedCode!, new DiagnosticContext (origin, diagnosticsEnabled: true, Context));
 
-			if (Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (method)) {
+			bool warnsOnReflection = Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (method);
+			if (warnsOnReflection && !skipWarningsForOverride) {
 				// ReflectionMethodBodyScanner handles more cases for data flow annotations
 				// so don't warn for those.
 				switch (dependencyKind) {
 				case DependencyKind.AttributeConstructor:
 				case DependencyKind.AttributeProperty:
-					return;
-
+					break;
 				default:
+					Context.LogWarning (origin, DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, method.GetDisplayName ());
 					break;
 				}
+			}
 
-				Context.LogWarning (origin, DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, method.GetDisplayName ());
+			// Warn about reflection access to compiler-generated code. This must happen even for virtual overrides,
+			// because it should include the virtual MoveNext method of state machines.
+			switch (dependencyKind) {
+			case DependencyKind.AccessedViaReflection:
+			case DependencyKind.DynamicallyAccessedMember:
+				if (ShouldWarnForReflectionAccessToCompilerGeneratedCode (method, warnsOnRUC, warnsOnReflection))
+					Context.LogWarning (origin, DiagnosticId.CompilerGeneratedMemberAccessedViaReflection, method.GetDisplayName ());
+				break;
```

Comment:
Same comment as above about the tricky interactions between override auto-suppression and the compiler-generated code. Probably worth the same comment (or reference one from the other).